### PR TITLE
client: Handle empty pathnames for `ceph_chownat()` and `ceph_statxat()`

### PR DESCRIPTION
**Pull request by anoopcs9**
man [fchownat(2)](https://www.man7.org/linux/man-pages/man2/fchownat.2.html) and man [statx(2)](https://www.man7.org/linux/man-pages/man2/statx.2.html) explains about the special case where empty _pathnames_ are valid with `AT_EMPTY_PATH` flag specified. But current client code crashes for _ceph_chownat()_ and _ceph_statxat()_ under this situation which should be gracefully handled as per the standards.

Fixes: https://tracker.ceph.com/issues/68189

Review with: `git show -w`

### CODE REVIEW

**Comment by anoopcs9** [comment]
`ceph_chownat()`:
```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007fea2757f6d3 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x00007fea27526c4e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fea2750e902 in __GI_abort () at abort.c:79
#4  0x00007fea266dabf0 in std::__glibcxx_assert_fail (file=file@entry=0x7fea27849680 "/usr/include/c++/14/string_view", line=line@entry=256, 
    function=function@entry=0x7fea27849580 "constexpr const std::basic_string_view<_CharT, _Traits>::value_type& std::basic_string_view<_CharT, _Traits>::operator[](size_type) const [with _CharT = char; _Traits = std::char_traits<char>; const_r"..., condition=condition@entry=0x7fea27854568 "__pos < this->_M_len") at ../../../../../libstdc++-v3/src/c++11/assert_fail.cc:41
#5  0x00007fea2774f6df in std::basic_string_view<char, std::char_traits<char> >::operator[] (this=<optimized out>, __pos=<optimized out>) at /usr/include/c++/14/string_view:254
#6  std::basic_string_view<char, std::char_traits<char> >::operator[] (this=<synthetic pointer>, __pos=0) at /usr/include/c++/14/string_view:254
#7  filepath::set_path (this=0x7fff29cd1e30, s=...) at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/include/filepath.h:101
#8  filepath::filepath (this=<optimized out>, s=<optimized out>, this=<optimized out>, s=<optimized out>) at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/include/filepath.h:94
#9  0x00007fea277a4764 in Client::chownat (this=0x32794a40, dirfd=10, relpath=0x4020d0 "", new_uid=65534, new_gid=65534, flags=0, perms=...)
    at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/client/Client.cc:8835
#10 0x00007fea2772a0c3 in ceph_chownat (cmount=<optimized out>, dirfd=<optimized out>, relpath=<optimized out>, uid=<optimized out>, gid=<optimized out>, flags=<optimized out>)
    at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/libcephfs.cc:1181
#11 0x00000000004013e1 in main ()
```


`ceph_statxat()`:
```
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007f6528fc26d3 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x00007f6528f69c4e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007f6528f51902 in __GI_abort () at abort.c:79
#4  0x00007f65280dabf0 in std::__glibcxx_assert_fail (file=file@entry=0x7f652928c680 "/usr/include/c++/14/string_view", line=line@entry=256, 
    function=function@entry=0x7f652928c580 "constexpr const std::basic_string_view<_CharT, _Traits>::value_type& std::basic_string_view<_CharT, _Traits>::operator[](size_type) const [with _CharT = char; _Traits = std::char_traits<char>; const_r"..., condition=condition@entry=0x7f6529297568 "__pos < this->_M_len") at ../../../../../libstdc++-v3/src/c++11/assert_fail.cc:41
#5  0x00007f65291926df in std::basic_string_view<char, std::char_traits<char> >::operator[] (this=<optimized out>, __pos=<optimized out>) at /usr/include/c++/14/string_view:254
#6  std::basic_string_view<char, std::char_traits<char> >::operator[] (this=<synthetic pointer>, __pos=0) at /usr/include/c++/14/string_view:254
#7  filepath::set_path (this=0x7ffe98822e10, s=...) at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/include/filepath.h:101
#8  filepath::filepath (this=<optimized out>, s=<optimized out>, this=<optimized out>, s=<optimized out>) at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/include/filepath.h:94
#9  0x00007f65291fad9e in Client::statxat (this=0x3bd71310, dirfd=10, relpath=0x4020d0 "", stx=0x7ffe98822ee0, perms=..., want=4095, flags=0)
    at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/client/Client.cc:11507
#10 0x00007f652916bd10 in ceph_statxat (cmount=<optimized out>, dirfd=<optimized out>, relpath=<optimized out>, stx=<optimized out>, want=<optimized out>, flags=<optimized out>)
    at /usr/src/debug/ceph-18.2.4-1.fc40.x86_64/src/libcephfs.cc:986
#11 0x0000000000401428 in main ()
```

**Review by dparmar18 — COMMENTED** [review]
looks good

**Comment by anoopcs9** [comment]
jenkins test make check

**Comment by anoopcs9** [comment]
jenkins test make check arm64

**Comment by anoopcs9** [comment]
jenkins test windows

**Comment by anoopcs9** [comment]
jenkins test make check arm64

**Comment by dparmar18** [comment]
jenkins test windows

**Review by dparmar18 — APPROVED** [review]
what i thought in my last approval was a comment, sorry!

**Review by chrisphoffman — CHANGES_REQUESTED** [review]
Can you also do a teuthology run to show that the linux code path passes?

File: src/test/libcephfs/test.cc [hunk 1]
```diff
@@ -2699,13 +2699,35 @@ TEST(LibCephFS, Statxat) {
...
   ASSERT_EQ(stx.stx_mode & S_IFMT, S_IFREG);
+  // test relative to dir with empty relpath
+#if defined(__linux__) && defined(AT_EMPTY_PATH)
+  int rel_file_fd = ceph_openat(cmount, fd, rel_file_name_1, O_RDONLY, 0);
+  ASSERT_LE(0, rel_file_fd);
+  ASSERT_EQ(ceph_statxat(cmount, rel_file_fd, "", &stx, AT_EMPTY_PATH, 0), 0);
+  ASSERT_EQ(stx.stx_mode & S_IFMT, S_IFREG);
+  ASSERT_EQ(0, ceph_close(cmount, rel_file_fd));
+#endif
+  ASSERT_EQ(0, ceph_close(cmount, fd));
```
[the remark this answers is not in the document]
Comment (anoopcs9) [reply src/test/libcephfs/test.cc]: Good catch, fixed in the updated version.

**Comment by anoopcs9** [comment]
> Can you also do a teuthology run to show that the linux code path passes?

Someone else will have to perform a teuthology run on my behalf I guess.

**Comment by anoopcs9** [comment]
jenkins test make check api

**Comment by anoopcs9** [comment]
jenkins test api

**Comment by petrutlucian94** [comment]
The Windows CI failure is legitimate, the test is failing on a clean Linux-only vstart environment:

```
ubuntu@ubuntu-ceph:/mnt/data/workspace/ceph_linux/build$ ./bin/ceph_test_libcephfs --gtest_filter="Lib
CephFS.Statxat"
Note: Google Test filter = LibCephFS.Statxat
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LibCephFS
[ RUN      ] LibCephFS.Statxat
/mnt/data/workspace/ceph_linux/src/test/libcephfs/test.cc:2706: Failure
Expected equality of these values:
  ceph_statxat(cmount, dir_fd, "", &stx, 0x1000, 0)
    Which is: -2
  0
[  FAILED  ] LibCephFS.Statxat (41 ms)
[----------] 1 test from LibCephFS (41 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (41 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LibCephFS.Statxat
```

The other Windows failure was legitimate as well as the other PR (forcefully merged) broke MDS, which now crashes whenever we attempt to do a mount: https://github.com/ceph/ceph/pull/58936#issuecomment-2378679026

**Comment by anoopcs9** [comment]
> I'm also seeing the same test failure as seen in: [#59936 (comment)](https://github.com/ceph/ceph/pull/59936#issuecomment-2379005576)
> 
> Can you track this down?

Oh..that was a silly mistake from my side mixing up the order of `flags` argument for `ceph_statxat()` in tests. PTAL.

**Review by chrisphoffman — CHANGES_REQUESTED** [review]
Running into an issue on linux:
```
$ ceph_test_libcephfs --gtest_filter=*Statxat
Note: Google Test filter = *Statxat
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LibCephFS
[ RUN      ] LibCephFS.Statxat
ceph/src/test/libcephfs/test.cc:2713: Failure
Expected equality of these values:
  ceph_statxat(cmount, dir_fd, "", &stx, 0, 0x1000)
    Which is: -22
  0
[  FAILED  ] LibCephFS.Statxat (60 ms)
[----------] 1 test from LibCephFS (60 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (60 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LibCephFS.Statxat

 1 FAILED TEST
```

Do you have a linux based machine to do local tests on? After you address the above,  can you run both tests you modified and post here the output passing?

**Comment by anoopcs9** [comment]
> Running into an issue on linux:
> 
> ```
> $ ceph_test_libcephfs --gtest_filter=*Statxat
> Note: Google Test filter = *Statxat
> [==========] Running 1 test from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 1 test from LibCephFS
> [ RUN      ] LibCephFS.Statxat
> ceph/src/test/libcephfs/test.cc:2713: Failure
> Expected equality of these values:
>   ceph_statxat(cmount, dir_fd, "", &stx, 0, 0x1000)
>     Which is: -22
>   0
> [  FAILED  ] LibCephFS.Statxat (60 ms)
> [----------] 1 test from LibCephFS (60 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (60 ms total)
> [  PASSED  ] 0 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] LibCephFS.Statxat
> 
>  1 FAILED TEST
> ```
> 
> Do you have a linux based machine to do local tests on? After you address the above, can you run both tests you modified and post here the output passing?

Fine, this time its `-22` which might point towards something internal preventing the added logic to work.

Alright I won't bother you hereafter until I make the tests work as expected. Thanks for running it through all this while.

**Comment by anoopcs9** [comment]
Update: It has to do with `CEPH_REQ_FLAG_MASK` which didn't allow the new `AT_EMPTY_PATH` flag from this change returning `CEPHFS_EINVAL`(22).

Now let me try running the tests as a double check.

**Comment by anoopcs9** [comment]
All set.
```
# git log --oneline origin/main..@
15fd978a59c (HEAD -> main) client: Gracefully handle empty pathname for statxat()
b3d9b271f91 libcephfs.h: Fix API documentation for ceph_statxat
5aa87e6dc3e client: Gracefully handle empty pathname for chownat()

# bin/ceph_test_libcephfs --gtest_filter="LibCephFS.Statxat"
Note: Google Test filter = LibCephFS.Statxat
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LibCephFS
[ RUN      ] LibCephFS.Statxat
[       OK ] LibCephFS.Statxat (673 ms)
[----------] 1 test from LibCephFS (673 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (673 ms total)
[  PASSED  ] 1 test.
```

File: src/libcephfs.cc [hunk 2]
```diff
@@ -982,7 +982,11 @@ extern "C" int ceph_statxat(struct ceph_mount_info *cmount, int dirfd, const cha
 {
   if (!cmount->is_mounted())
     return -CEPHFS_ENOTCONN;
+#if defined(__linux__) && defined(AT_EMPTY_PATH)
+  if (flags & ~(CEPH_REQ_FLAG_MASK|AT_EMPTY_PATH))
```
Comment (dparmar18) [inline src/libcephfs.cc:986]: How about defining this flag in `src/include/cephfs/ceph_ll_client.h`? Not sure if it makes sense but couldn't stop myself from commenting

Comment (anoopcs9) [reply src/libcephfs.cc:986]: Did you mean to define `CEPH_REQ_FLAG_MASK|AT_EMPTY_PATH` as another inside _src/include/cephfs/ceph_ll_client.h_?

**Comment by anoopcs9** [comment]
jenkins test api

**Comment by anoopcs9** [comment]
jenkins test make check

File: src/libcephfs.cc (hunk 2 above)
Comment (dparmar18) [reply src/libcephfs.cc:986]: Yes, as a new statx interface flag

Comment (anoopcs9) [reply src/libcephfs.cc]: Let me know if the updated version goes along with your suggestion.

File: src/include/cephfs/ceph_ll_client.h [hunk 3]
```diff
@@ -110,6 +110,9 @@ struct ceph_statx {
  * others in the future, we disallow setting any that aren't recognized.
  */
 #define CEPH_REQ_FLAG_MASK		(AT_SYMLINK_NOFOLLOW|AT_STATX_DONT_SYNC)
+#if defined(__linux__) && defined(AT_EMPTY_PATH)
+#define CEPH_REQ_FLAG_MASK_EXT		(CEPH_REQ_FLAG_MASK|AT_EMPTY_PATH)
```
[the remark this answers is not in the document]
Comment (anoopcs9) [reply src/include/cephfs/ceph_ll_client.h:114]: Yes, but I'm open to other naming suggestions.

**Review by chrisphoffman — COMMENTED** [review]
Tests look good!

Since we're adding support for `AT_EMPTY_PATH` for some calls, I recommend we look and add support to all calls that normally handle it. That way we won't have to circle back to this in the future.

For example:
`link`, `stat`, `lstat`  etc

File: src/include/cephfs/ceph_ll_client.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (dparmar18) [reply src/include/cephfs/ceph_ll_client.h:114]: How about just renaming this to `CEPH_AT_EMPTY_PATH`? This enables the bits of both the flags but I guess we're more interested in `AT_EMPTY_PATH` right?

[the remark this answers is not in the document]
Comment (anoopcs9) [reply src/include/cephfs/ceph_ll_client.h:114]: > Can you explain further why `CEPH_REQ_FLAG_MASK_EXT` is needed versus including `AT_EMPTY_PATH` via `CEPH_REQ_FLAG_MASK`?

Two reasons:
- `AT_EMPTY_PATH` is an optional flag available with just Linux as part of **__GNU_SOURCE_** macro(see  [feature_test_macros(7)](https://www.man7.org/linux/man-pages/man7/feature_test_macros.7.html)). So this has to be conditionally defined.
- `CEPH_REQ_FLAG_MASK` is already used as a filter at many other places, which I prefer not to touch with where `AT_EMPTY_PATH` might not be valid.

> I'm wondering if we could minimize the amount of defines and ifdefs.

I know but this is how it is 😐 .

> How about just renaming this to `CEPH_AT_EMPTY_PATH`? This enables the bits of both the flags but I guess we're more interested in `AT_EMPTY_PATH` right?

I agree, I'll update.

**Review by anoopcs9 — COMMENTED** [review]
> Since we're adding support for `AT_EMPTY_PATH` for some calls, I recommend we look and add support to all calls that normally handle it. That way we won't have to circle back to this in the future.

This was briefly touched on previously [here](https://github.com/ceph/ceph/pull/59503#discussion_r1768035659) and the current PR was a followup to that discussion.

> For example: `link`, `stat`, `lstat` etc

IIUC, `AT_EMPTY_PATH` flag is only relevant with _*at()_ calls and not others. With an implementation missing for _linkat()_ I don't think we are left with any APIs accepting `AT_EMPTY_PATH`. _faccessat()_ and _utimensat()_ are comparatively recent additions(since Linux 5.8) to honour this flag which I prefer to be considered after a while(as we might still have systems with Linux < 5.8).

**Review by dparmar18 — APPROVED** [review]
LGTM

**Comment by chrisphoffman** [comment]
> > Since we're adding support for `AT_EMPTY_PATH` for some calls, I recommend we look and add support to all calls that normally handle it. That way we won't have to circle back to this in the future.
> 
> This was briefly touched on previously [here](https://github.com/ceph/ceph/pull/59503#discussion_r1768035659) and the current PR was a followup to that discussion.
> 
I missed that, wanted to make sure we followed up on remaining calls. Thanks!

> > For example: `link`, `stat`, `lstat` etc
> 
> IIUC, `AT_EMPTY_PATH` flag is only relevant with _*at()_ calls and not others. With an implementation missing for _linkat()_ I don't think we are left with any APIs accepting `AT_EMPTY_PATH`. _faccessat()_ and _utimensat()_ are comparatively recent additions(since Linux 5.8) to honour this flag which I prefer to be considered after a while(as we might still have systems with Linux < 5.8).

I came up with those examples above from the man pages. I agree it seems like it should only apply at _*at()_ calls.

That being said, I'm still a bit confused if this is actually the case. There's `statx` that accept flags and then call the _*at()_ functions directly.

See:
https://github.com/ceph/ceph/blob/4598444db607e50f0295922492053a7393dc8de0/src/client/Client.cc#L8583-L8588

See the man page here to see `AT_EMPTY_PATH`:
https://man7.org/linux/man-pages/man2/statx.2.html

nit: While that shows there's no code that needs to be changed, I recommend changes in libcephfs.h show AT_EMPTY_PATH does work now for `statx`.
https://github.com/ceph/ceph/blob/4598444db607e50f0295922492053a7393dc8de0/src/include/cephfs/libcephfs.h#L953

**Review by chrisphoffman — APPROVED** [review]
LGTM- thanks for all your work and discussion!

**Comment by anoopcs9** [comment]
> That being said, I'm still a bit confused if this is actually the case. There's `statx` that accept flags and then call the _*at()_ functions directly.
> 
> See:
> 
> https://github.com/ceph/ceph/blob/4598444db607e50f0295922492053a7393dc8de0/src/client/Client.cc#L8583-L8588
> 
> See the man page here to see `AT_EMPTY_PATH`: https://man7.org/linux/man-pages/man2/statx.2.html

tl;dr
_ceph_statx()_ doesn't comply with the standard _statx()_ because it doesn't accept the required **_dirfd_** argument instead always operates at `AT_FDCWD` and thus we have _ceph_statxat()_ to handle this case.

longer version
As you already know we have `struct stat` and `struct statx` these days although I think `struct statx` is preferred. Normal calls like _stat()_, _fstat()_, _lstat()_ and _fstatat()_ operates with `struct stat`. Among these only _fstatat()_ is relevant for `AT_EMPTY_PATH` flag. But ceph doesn't have an equivalent _*at()_ API designed to operate with `struct stat`. Instead we have two variants to work with `struct statx` namely _ceph_statx()_ and _ceph_statxat()_. It is the latter which accepts a **_dirfd_** argument. That's why it make sense to honour `AT_EMPTY_PATH` flag only for _ceph_statxat()_ and not _ceph_statx()_.

Given that, I think it would make more sense to reference [statx(2)](https://man7.org/linux/man-pages/man2/statx.2.html) than [fstatat(2)](https://www.man7.org/linux/man-pages/man2/fstatat.2.html) in the commit message/PR description. Let me know if I should update once more.

> nit: While that shows there's no code that needs to be changed, I recommend changes in libcephfs.h show AT_EMPTY_PATH does work now for `statx`.
> 
> https://github.com/ceph/ceph/blob/4598444db607e50f0295922492053a7393dc8de0/src/include/cephfs/libcephfs.h#L953

I think the earlier explanation invalidates the above suggestion.

**Comment by anoopcs9** [comment]
I've updated commit message and PR description to link [statx(2)](https://www.man7.org/linux/man-pages/man2/statx.2.html) instead of [fstatat(2)](https://www.man7.org/linux/man-pages/man2/fstatat.2.html).

**Comment by vshankar** [comment]
> @vshankar can I have the final blessings here?

@mchangir is involved in running tests for sometime (I've been tied up with some cephalocon stuff).

**Comment by mchangir** [comment]
This PR is under test in [https://tracker.ceph.com/issues/69069](https://tracker.ceph.com/issues/69069).

**Comment by anoopcs9** [comment]
> This PR is under test in https://tracker.ceph.com/issues/69069.

@mchangir Where are we with the teuthology runs? Is this good to go in now?

**Comment by mchangir** [comment]
[wiki section for qa run](https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-mchangir-testing-mon-caps-main-debug)

**Comment by vshankar** [comment]
> > > This PR is under test in https://tracker.ceph.com/issues/69069.
> > 
> > 
> > @mchangir Where are we with the teuthology runs? Is this good to go in now?
> 
> looks okay ... I've just finished reviewing the QA run Venky should be able to ack it in a day or two

If the test run passes, please go ahead and merge. @mchangir
